algorithms/kernel/k_nearest_neighbors/oneapi/bf_knn_classification_predict_kernel_ucapi_impl.i

```diff
@@ -144,145 +114,119 @@ Status KNNClassificationPredictKernelUCAPI<algorithmFpType>::
             {
-                QuickSelectIndexed::select(partialDistances, partialCategories, rndSeq, dataBlockSize, nK, curProbeBlockSize,
-                                            nK * partCount, nK * nParts, selectResult, &st);  DAAL_CHECK_STATUS_VAR(st);
+                // Select nK closest neighbors from [curProbeBlockSize]x[nK * partCount] buffers with partial results
+                selector->selectLabels(partialDistances, partialCategories, nK, curProbeBlockSize, nK * partCount, nK * nParts, nK * nParts,
+                                       selectResult, &st);
+                DAAL_CHECK_STATUS_VAR(st);
                 partCount = 0;
             }
+            DAAL_CHECK_STATUS_VAR(labels->releaseBlockOfRows(labelRows));
+            DAAL_CHECK_STATUS_VAR(points->releaseBlockOfRows(dataRows));
         }
-        if(partCount > 0)
-        {
-            QuickSelectIndexed::select(partialDistances, partialCategories, rndSeq, dataBlockSize, nK, curProbeBlockSize,
-                                        nK * partCount, nK * nParts, selectResult, &st);  DAAL_CHECK_STATUS_VAR(st);
-        }
-        RadixSort::sort(selectResult.indices, sorted,  radix_buffer, curProbeBlockSize, nK, nK, &st); DAAL_CHECK_STATUS_VAR(st);
+        if (partCount > 0)
         {
-            DAAL_ITTNOTIFY_SCOPED_TASK(compute.winners);
-            computeWinners(context, find_winners, sorted, classes, curProbeBlockSize, nK, &st); DAAL_CHECK_STATUS_VAR(st);
+            // Select nK closest neighbors from buffers with remaining partial results
+            selector->selectLabels(partialDistances, partialCategories, nK, curProbeBlockSize, nK * partCount, nK * nParts, nK * nParts, selectResult,
+                                   &st);
+            DAAL_CHECK_STATUS_VAR(st);
         }
+        // sort labels of closest neighbors
+        RadixSort::sort(selectResult.indices, sorted, radix_buffer, curProbeBlockSize, nK, nK, &st);
+        DAAL_CHECK_STATUS_VAR(st);
+        // search for maximum occurrence label
+        computeWinners(context, find_winners, sorted, classes, curProbeBlockSize, nK, &st);
+        DAAL_CHECK_STATUS_VAR(st);
+        // copy final results
         {
             DAAL_ITTNOTIFY_SCOPED_TASK(compute.copy_results);
             BlockDescriptor<algorithmFpType> resultBlock;
             DAAL_CHECK_STATUS_VAR(y->getBlockOfRows(0, nProbeRows, writeOnly, resultBlock));
-            auto classesRows = classes.template get<int>().toHost(ReadWriteMode::readOnly);
-            auto res = resultBlock.getBuffer().toHost(ReadWriteMode::writeOnly);
-                for(int i = 0; i < curProbeBlockSize; i++)
-                {
-                    res.get()[pfirst + i] = classesRows.get()[i];
-                }
+            auto classesRows = classes.template get<int>().toHost(ReadWriteMode::readOnly, &st);
+            DAAL_CHECK_STATUS_VAR(st);
+            auto res = resultBlock.getBuffer().toHost(ReadWriteMode::writeOnly, &st);
```

Comment:
Why you are performing copy operation from host? And why do you need to copy results instead of calculating them in `y` table?